src/InsertCalorimeter_geo.cpp

```diff
@@ -0,0 +1,207 @@
+
+  // Getting thickness of backplate
+  auto backplate_thickness = dd4hep::getAttrOrDefault<double>(detElem, _Unicode(backplate_thickness), 1.61 * cm);
+
+  // Function that returns a linearly interpolated hole radius at a given z
+  auto get_hole_radius = [hole_radius_initial, hole_radius_final, length, backplate_thickness](double z_pos) {
+    /*
+        radius = (hole_radius_final - hole_radius_initial)/(length) * z_pos +  hole_radius_initial
+        Treats the beginning of the insert as z = 0
+        At z = 0 (beginning of first layer), hole radius is hole_radius_initial
+        The radius is hole_radius_final at the beginning of the backplate,
+          i.e. z = length - backplate_thickness
+    */
+    double slope = (hole_radius_final - hole_radius_initial) / (length - backplate_thickness);
+    return slope * z_pos + hole_radius_initial;
+  };
+
+  // Defining envelope
+  Box envelope(width / 2.0, height / 2.0, length / 2.0);
+
+  /*
+    Hole initial x-position in local coordinate system
+    x = 0 in local coordinate system is x = HcalEndcapPInsert_xposition in global coordinate system
+    Hole starts at x = -7.29 cm in global system
+  */
+  const double initial_hole_x = -7.29 - pos.x();
+
+  // Keeps track of the hole's x location for each layer
+  double hole_x_tracker = initial_hole_x;
+
+  // Keeps track of the z location as we move longiduinally through the insert
+  // Will use this tracking variable as input to get_hole_radius
+  double z_distance_traversed = 0.;
+
+  /*
+    Cutting out the hole from the envelope
+    Cut out hole layer-by-layer to get exact desired shape
+    Subtracting a cone is difficult since the cone would have to be angled towards the negative x direction
+  */
+
+  // Looping through all the different layer sections (W/Sc, Steel/Sc, backplate)
+  for (xml_coll_t c(detElem, _U(layer)); c; ++c) {
+    xml_comp_t x_layer         = c;
+    int        repeat          = x_layer.repeat();
+    double     layer_thickness = x_layer.thickness();
+
+    // Looping through the number of repeated layers in each section
+    for (int ilayer = 0; ilayer < repeat; ilayer++) {
+      const double hole_radius = get_hole_radius(z_distance_traversed);
+      Tube         layer_hole(0., hole_radius, layer_thickness / 2.);
+
+      /*
+        X-Position:
+        The hole starts at x = 2.71 cm with respect to local coordinate system.
+        The hole shifts by -.0569 cm with each layer
+
+        Z-Position:
+        -length / 2. is front of insert,
+        +z_distance_traversed goes to the front of each layer,
+        +layer_thickness / 2. is the half-length of a layer
+          (i.e. where to put the cutout shape)
+      */
+
+      SubtractionSolid envelope_with_hole(
+          envelope, layer_hole, Position(hole_x_tracker, 0., (-length + layer_thickness) / 2. + z_distance_traversed));
+
+      // Removing the hole layer by layer from envelope
+      envelope = envelope_with_hole;
```

Comment:
I think we should avoid these serial subtraction solids. They are notoriously inefficient since they can't be optimized and always incur a geometry lookup on all components. Can we use a polycone here, which at least makes this a bit more efficient?

Reply:
I tried looking into this yesterday evening but I'm pretty lost on how to implement this. Do you have any references/guides on how to use a polycone? We can also discuss this during the office hours tomorrow.

Reply:
This may be useful: https://github.com/eic/ip6/blob/533d83a4e7c993b9a3b729ba20c89a7b01c41e27/src/IP6BeamPipe.cpp#L143 (but it's also overkill for this here). Essentially you specify rmin/rmax pairs for fixed z planes. So, in your case you'd fill a `std::vector<double> rmax, rmin, z` in the loop like you are already doing, then use those once to create a `Polycone(0, 2.0 * M_PI, rmin, rmax, z)` that uses those rmin/rmax pairs at z planes. If necessary, you can pick rmin always zero, only vary rmax, then finally do a single subtraction from the original `envelope`. It would change `repeat` serial subtractions into just one polycone subtraction from `envelope`.

Reply:
Hmm it seems Polycone doesn't take x and y positions for each segment. Each layer of the insert has a hole with a different radius and at a different x position. Is there a way to account for this easily? Maybe I could make a union of a bunch of circles at different positions and then subtract that? But I'm guessing that causes the same optimization issue as before

Reply:
For documentation: Implemented an assembly instead of an envelope. Polycone doesn't support shifts in x and y. Need twisted cone shape but doesn't exist and time consuming to implement. Using assembly is more efficient than serial subtraction solids.